plugins/nordic-hid/fu-nordic-hid-cfg-channel.c

```diff
@@ -436,61 +501,226 @@ fu_nordic_hid_cfg_channel_add_peers(FuNordicHidCfgChannel *self, GError **error)
+						NULL,
+						NULL,
+						CONFIG_STATUS_GET_PEERS_CACHE,
+						NULL,
+						0,
+						error))
+		return FALSE;
+
+	if (fu_nordic_hid_cfg_channel_cmd_receive(self,
+						  CONFIG_STATUS_DISCONNECTED,
+						  res,
+						  &error_local)) {
+		/* Configuration channel peers cache not supported. */
+		return TRUE;
+	}
+
+	/* Configuration channel peer caching available */
+	if (!fu_nordic_hid_cfg_channel_cmd_receive(self, CONFIG_STATUS_SUCCESS, res, error))
+		return FALSE;
+
+	if (!fu_memcpy_safe(peers_cache,
+			    PEERS_CACHE_LEN,
+			    0,
+			    res->data,
+			    PEERS_CACHE_LEN,
+			    0,
+			    PEERS_CACHE_LEN,
+			    error))
+		return FALSE;
+
+	*cmd_supported = TRUE;
+
+	return TRUE;
+}
+
+static gboolean
+fu_nordic_hid_cfg_channel_update_peers(FuNordicHidCfgChannel *self,
+				       GError **error,
+				       guint8 peers_cache[PEERS_CACHE_LEN])
+{
+	g_autoptr(GError) error_local = NULL;
+	gboolean peers_supported = FALSE;
+	guint8 peer_id;
+	guint cnt = 0;
+
+	if (!fu_nordic_hid_cfg_channel_index_peers_cmd(self, error, &peers_supported))
+		return FALSE;
+
+	if (!peers_supported)
+		return TRUE;
+
+	/* a device that does not support peers caching, would drop all of the peers because it
+	 * cannot determine if the previously discovered peer is still connected.
+	 */
+	fu_nordic_hid_cfg_channel_remove_disconnected_peers(self, peers_cache);
+
 	while (cnt++ <= 0xFF) {
-		g_autoptr(FuNordicHidCfgChannel) peer = NULL;
-		g_autoptr(GError) error_peer = NULL;
+		if (!fu_nordic_hid_cfg_channel_get_next_peer_id_cmd(self, error, &peer_id))
+			return FALSE;
 
-		if (!fu_nordic_hid_cfg_channel_cmd_send(self,
-							NULL,
-							NULL,
-							CONFIG_STATUS_GET_PEER,
-							NULL,
-							0,
-							error))
+		/* end of the list */
+		if (peer_id == INVALID_PEER_ID)
+			break;
+
+		g_debug("detected peer: 0x%02x", peer_id);
+
+		if (peers_cache == NULL) {
+			/* allow to properly discover dongles without peers cache support. */
+			fu_nordic_hid_cfg_channel_add_peer(self, peer_id);
+		} else {
+			guint8 peer_cache_idx = peer_id - 1;
+
+			if (self->peers_cache[peer_cache_idx] != peers_cache[peer_cache_idx] &&
+			    is_cached_peer_connected(peers_cache[peer_cache_idx])) {
+				fu_nordic_hid_cfg_channel_remove_peer(self, peer_id);
+				fu_nordic_hid_cfg_channel_add_peer(self, peer_id);
+				self->peers_cache[peer_cache_idx] = peers_cache[peer_cache_idx];
+			}
+		}
+	}
+
+	if (peer_id != INVALID_PEER_ID) {
+		g_set_error_literal(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_BROKEN_PIPE,
+				    "too many peers detected");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_nordic_hid_cfg_channel_setup_peers(FuNordicHidCfgChannel *self, GError **error)
+{
+	gboolean peers_supported = FALSE;
+	gboolean peers_cache_supported = FALSE;
+	guint8 peers_cache[PEERS_CACHE_LEN] = {0x00};
+
+	if (self->peer_id != 0)
+		/* device connected through dongle cannot support peers. */
+		return TRUE;
```

Comment:
needs braces

Reply:
Thanks, I missed this while adding a comment